### PR TITLE
improved error messages for inexistent userscripts

### LINKED ISSUE #1940: Check that userscript exists before attempting to execute it
If a userscript is invoked by the `:spawn --userscript myscript` command, but `myscript` does not exist in the userscripts directory, qutebrowser gives the uninformative message:

    Error while spawning userscript: The process failed to start.

The log output is similarly opaque. The log output does make it clear that qutebrowser simply assumes the existence of the userscript and attempts to execute it.

It would be more useful if qutebrowser checked that the userscript exists in the userscript directory before attempting to execute it. If the userscript cannot be found qutebrowser could provide a specific, more helpful, error message like:

    Userscript 'myscript' not found in userscript directory 'path/to/userscript/directory'

### PR DESCRIPTION
**Pull request by HolySmoke86**
fixes #1940

### CODE REVIEW

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -1122,7 +1122,7 @@ def _run_userscript(self, cmd, *args, verbose=False):
         try:
             userscripts.run_async(tab, cmd, *args, win_id=self._win_id,
                                   env=env, verbose=verbose)
-        except userscripts.UnsupportedError as e:
+        except (userscripts.UnsupportedError, userscripts.NotFoundError) as e:
```
Comment (HolySmoke86) [inline qutebrowser/browser/commands.py]: @The-Compiler Is this style okay or would you prefer a common userscript exception base?

File: qutebrowser/commands/userscripts.py [hunk 2]
```diff
@@ -315,6 +315,29 @@ def prepare_run(self, *args, **kwargs):
...
+    """
+
+    def __init__(self, script_name, paths=None):
+        super().__init__()
+        self.script_name = script_name
+        self.paths = paths
+        self.message = "Userscript '" + script_name + "' not found"
+        if paths:
+            self.message += " in userscript directory '" + \
+                            ("' or '".join(paths)) + "'"
```
Comment (The-Compiler) [inline qutebrowser/commands/userscripts.py]: I'd prefer:

```python
            self.message += "in userscript directories {}".format(
                ', '.join(repr(path) for path in paths))
```

This uses `, ` as delimiter which seems more appropriate to me, says "directories", and makes the whole thing more readable IMHO :wink:

File: qutebrowser/commands/userscripts.py [hunk 3]
```diff
@@ -323,12 +346,41 @@ def __str__(self):
...
+
+    Args:
+        cmd: The command to look for.
+
+    Returns:
+        A path to the userscript.
+    """
+    cmd_path = os.path.join(standarddir.data(), "userscripts", cmd)
+    if os.path.exists(cmd_path):
+        return cmd_path
```
Comment (The-Compiler) [inline qutebrowser/commands/userscripts.py]: Why is this needed? Don't the lines below with the first entry in `directories` do the same?

File: tests/end2end/features/spawn.feature [hunk 4]
```diff
@@ -10,7 +10,11 @@ Feature: :spawn
 
     Scenario: Starting a userscript which doesn't exist
         When I run :spawn -u this_does_not_exist
-        Then the error "Error while spawning userscript: The process failed to start." should be shown
+        Then regex "Userscript 'this_does_not_exist' not found in userscript directory '(.*)'( or '(.*)')*" should be logged
```
Comment (The-Compiler) [inline tests/end2end/features/spawn.feature]: This seems too specific, I think "... in userscript directory *" (not as regex) is good enough.

File: qutebrowser/browser/commands.py (hunk 1 above)
Comment (HolySmoke86) [inline qutebrowser/browser/commands.py]: I'll add it then, better sooner than later

File: qutebrowser/commands/userscripts.py [hunk 5]
```diff
@@ -315,6 +315,29 @@ def prepare_run(self, *args, **kwargs):
...
 
 
+class NotFoundError(Exception):
+
+    """Raised when spawning a userscript that doesn't exist.
+
+    Attributes:
+        script_name: name of the userscript as called
+        paths: path names that were searched for the userscript
+        message: descriptive message string of the error
```
Comment (HolySmoke86) [inline qutebrowser/commands/userscripts.py]: My (possibly bad) habit of caching... I'll move it to `str()` in combination with the string formatting

File: qutebrowser/commands/userscripts.py [hunk 6]
```diff
@@ -315,6 +315,29 @@ def prepare_run(self, *args, **kwargs):
...
+    """
+
+    def __init__(self, script_name, paths=None):
+        super().__init__()
+        self.script_name = script_name
+        self.paths = paths
+        self.message = "Userscript '" + script_name + "' not found"
```
Comment (HolySmoke86) [inline qutebrowser/commands/userscripts.py]: That is awesome, I need to habitualize this string.format thing

File: tests/end2end/features/spawn.feature (hunk 4 above)
Comment (HolySmoke86) [inline tests/end2end/features/spawn.feature]: Ah, I didn't know `*` worked in those. I tried `(.*)` with the `message` `should be shown` clause

**Comment by The-Compiler** [comment]
Thanks!
